File: backend/app/schemas/grouping.py
"""
Pydantic schemas for UW Model File Grouping & Data Extraction API.
"""

from typing import Any

from pydantic import BaseModel, Field


class GroupSummary(BaseModel):
    """Summary of a single file group."""

    group_name: str
    file_count: int = 0
    structural_overlap: float = 1.0
    era: str = ""
    sub_variant_count: int = 0


class DiscoveryResponse(BaseModel):
    """Response from Phase 1 discovery."""

    total_scanned: int
    candidates_accepted: int
    candidates_skipped: int
    duplicates_removed: int
    batch_info: dict[str, Any] | None = None


class FingerprintResponse(BaseModel):
    """Response from Phase 2.1 fingerprinting."""

    total_fingerprinted: int
    populated: int
    sparse: int
    empty_templates: int


class GroupListResponse(BaseModel):
    """Response listing all groups."""

    groups: list[GroupSummary]
    total_groups: int
    total_ungrouped: int
    total_empty_templates: int


class GroupDetailResponse(BaseModel):
    """Detailed response for a single group."""

    group_name: str
    files: list[dict[str, Any]] = Field(default_factory=list)
    structural_overlap: float = 1.0
    era: str = ""
    sub_variants: list[str] = Field(default_factory=list)
    variances: dict[str, Any] = Field(default_factory=dict)


class ReferenceMappingResponse(BaseModel):
    """Response from Phase 3 reference mapping."""

    groups_mapped: int
    total_fields_mapped: int
    total_fields_unmapped: int
    per_group: dict[str, dict[str, Any]] = Field(default_factory=dict)


class ConflictCheckResponse(BaseModel):
    """Response from Phase 4.1 conflict check."""

    groups_with_conflicts: int
    total_conflicts: int
    conflicts: dict[str, list[dict[str, Any]]] = Field(default_factory=dict)


class GroupExtractionRequest(BaseModel):
    """Request to extract data from a group."""

    dry_run: bool = Field(
        default=True, description="If True, produces report without DB writes"
    )


class GroupExtractionResponse(BaseModel):
    """Response from Phase 4.2 group extraction."""

    group_name: str
    dry_run: bool
    files_processed: int
    files_failed: int
    total_values: int
    started_at: str = ""
    completed_at: str = ""


class PipelineStatusResponse(BaseModel):
    """Full pipeline status response."""

    data_dir: str
    phases: dict[str, str]
    stats: dict[str, int]
    created_at: str = ""
    updated_at: str = ""

File: src/types/grouping.ts
/**
 * Types for the UW Model File Grouping pipeline — groups discovered UW model files
 * by structural similarity and manages per-group extraction, approval, and validation.
 */

// ---------------------------------------------------------------------------
// Pipeline status
// ---------------------------------------------------------------------------

export interface PipelineStatus {
  data_dir: string;
  phases: Record<string, string>; // phase name -> completion timestamp or empty
  stats: Record<string, number>; // aggregate counts
  created_at: string;
  updated_at: string;
}

// ---------------------------------------------------------------------------
// Group summary (list view)
// ---------------------------------------------------------------------------

export interface GroupSummary {
  group_name: string;
  file_count: number;
  structural_overlap: number;
  era: string;
  sub_variant_count: number;
}

// ---------------------------------------------------------------------------
// Group detail
// ---------------------------------------------------------------------------

export interface GroupDetail {
  group_name: string;
  files: Array<Record<string, unknown>>;
  structural_overlap: number;
  era: string;
  sub_variants: string[];
  variances: Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// Phase responses
// ---------------------------------------------------------------------------

/** POST /extraction/grouping/discover */
export interface DiscoveryResponse {
  total_scanned: number;
  candidates_accepted: number;
  candidates_skipped: number;
  duplicates_removed: number;
  batch_info?: Record<string, unknown>;
}

/** POST /extraction/grouping/fingerprint */
export interface FingerprintResponse {
  total_fingerprinted: number;
  populated: number;
  sparse: number;
  empty_templates: number;
}

/** GET /extraction/grouping/groups */
export interface GroupListResponse {
  groups: GroupSummary[];
  total_groups: number;
  total_ungrouped: number;
  total_empty_templates: number;
}

/** POST /extraction/grouping/reference-map */
export interface ReferenceMappingResponse {
  groups_mapped: number;
  total_fields_mapped: number;
  total_fields_unmapped: number;
  per_group: Record<string, Record<string, unknown>>;
}

/** POST /extraction/grouping/conflict-check */
export interface ConflictCheckResponse {
  groups_with_conflicts: number;
  total_conflicts: number;
  conflicts: Record<string, Array<Record<string, unknown>>>;
}

// ---------------------------------------------------------------------------
// Extraction responses
// ---------------------------------------------------------------------------

/** POST /extraction/grouping/extract/{name} */
export interface GroupExtractionResponse {
  group_name: string;
  dry_run: boolean;
  files_processed: number;
  files_failed: number;
  total_values: number;
  started_at: string;
  completed_at: string;
}

/** POST /extraction/grouping/approve/{name} */
export interface GroupApprovalResponse {
  group_name: string;
  approved: boolean;
  message: string;
}

/** POST /extraction/grouping/extract-batch */
export interface BatchExtractionResponse {
  groups_processed: number;
  groups_failed: number;
  total_files: number;
  total_values: number;
  per_group: Record<string, GroupExtractionResponse>;
}

// ---------------------------------------------------------------------------
// Validation response
// ---------------------------------------------------------------------------

/** POST /extraction/grouping/validate */
export interface ValidationResponse {
  groups_validated: number;
  groups_passed: number;
  groups_failed: number;
  per_group: Record<string, Record<string, unknown>>;
}
